Sming/system/flashmem.c

```diff
@@ -1,102 +1,205 @@
+		if(FLASHMEM_READ(tmpdata, addr_aligned, blksize) != blksize)
+			return 0;
+		while(remain && rest < blksize)
+		{
+			tmpdata[rest++] = *pfrom++;
+			--remain;
+		}
+
+		// Write the unit
+		uint32_t written = FLASHMEM_WRITE(tmpdata, addr_aligned, blksize);
+		if(written != blksize)
+			return written;
+
+		if (remain == 0)
+			return size;
+
+		toaddr = addr_aligned + blksize;
+	}
+
+	// The start address is now a multiple of blksize
+	// Compute how many bytes we can write as multiples of blksize
+	uint32_t rest = remain & blkmask;
+	remain &= ~blkmask;
+	// Program the blocks now
+	while(remain)
+	{
+		unsigned count = min(remain, sizeof(tmpdata));
+		memcpy(tmpdata, pfrom, count);
+		uint32_t written = FLASHMEM_WRITE(tmpdata, toaddr, count);
+		remain -= written;
+		if(written != count)
+			return size - remain;
+		toaddr += count;
+		pfrom += count;
+	}
+
+	// And the final part of a block if needed
+	if(rest)
+	{
+		if(FLASHMEM_READ(tmpdata, toaddr, blksize) != blksize)
+			return size - remain;
+		unsigned i;
+		for(i = 0; i < rest; ++i)
+			tmpdata[i] = *pfrom++;
+		uint32_t written = FLASHMEM_WRITE(tmpdata, toaddr, blksize);
+		remain -= written;
+		if(written != blksize)
+			return size - remain;
+	}
+
+	return size;
+}
+
+uint32_t flashmem_read(void* to, uint32_t fromaddr, uint32_t size)
+{
+	if(IS_ALIGNED(to) && IS_ALIGNED(fromaddr) && IS_ALIGNED(size))
+		return FLASHMEM_READ(to, fromaddr, size);
+
+	const uint32_t blksize = INTERNAL_FLASH_READ_UNIT_SIZE;
+	const uint32_t blkmask = INTERNAL_FLASH_READ_UNIT_SIZE - 1;
+
+	__aligned char tmpdata[FLASH_BUFFERS * INTERNAL_FLASH_READ_UNIT_SIZE];
+	uint8_t* pto = (uint8_t*)to;
+	uint32_t remain = size;
+
+	// Align the start
+	if(fromaddr & blkmask)
+	{
+		uint32_t rest = fromaddr & blkmask;
+		uint32_t addr_aligned = fromaddr & ~blkmask; // this is the actual aligned address
+		if (FLASHMEM_READ(tmpdata, addr_aligned, blksize) != blksize)
+			return 0;
+		// memcpy(pto, &tmpdata[rest], std::min(blksize - rest, remain))
+		while(remain && rest < blksize)
+		{
+			*pto++ = tmpdata[rest++];
+			--remain;
+		}
+		if (remain == 0)
+			return size;
+		fromaddr = addr_aligned + blksize;
+	}
+
+	// The start address is now a multiple of blksize
+	// Compute how many bytes we can read as multiples of blksize
+	uint32_t rest = remain & blkmask;
+	remain &= ~blkmask;
+	// Read the blocks now
+	while(remain)
+	{
+		unsigned count = min(remain, sizeof(tmpdata));
+		uint32_t read = FLASHMEM_READ(tmpdata, fromaddr, count);
+		memcpy(pto, tmpdata, read);
+		remain -= read;
+		if(read != count)
+			return size - remain;
+		fromaddr += count;
+		pto += count;
+	}
+
+	// And the final part of a block if needed
+	if(rest)
+	{
+		if(FLASHMEM_READ(tmpdata, fromaddr, blksize) != blksize)
+			return size - remain;
+		unsigned i;
+		for(i = 0; i < rest; ++i)
+			*pto++ = tmpdata[i];
+	}
+
+	return size;
+}
+
+bool flashmem_erase_sector(uint32_t sector_id)
+{
+	WDT_FEED();
```

Comment:
Is `WDT_FEED()` really needed? I am just curious if you have experienced any issues with flash secor erasing causing the software watchdog to kick in?

Reply:
This question applies also to the other new lines where WDT_FEED() is called.

Reply:
I haven't had any issues, but then this was always in the code as `WRITE_PERI_REG(0x60000914, 0x73);`. I had to dig around to find out what that actually did, turned out there was a useful definition for it.

Reply:
I think the main risk from watchdog timeouts is with `spiffs_format`. Erase/write operations for normal file access would normally be much shorter, unlikely enough to trip a watchdog.

Reply:
My previous statement about 3ms, ignore please, it's wrong! I spent a little more time on this today and put timing checks on `fileWrite`. Writes take an average of about 10us per byte, reads no more than 1us. Unlikely to write more than 10K in one operation (RAM limitations) so, with overheads, maybe 150ms for a fileWrite, 15ms for a fileRead.

Given these figures, the WDT seems un-necessary. If we wanted to remove them we'd need to take a closer look at `spiffs_format` and `spiffs_check`. Everything else is done with buffering and callbacks so should be fine. For now, probably easier to leave them in.